File: services/core/internal/admin/administrator.go
// SHIP-147: what an administrator account is, and what an administrator session is.
//
// The types only. What they mean is in three other places and each is deliberately separate:
// credentials.go signs one in, adminauth.go verifies one on a request, and permissions.go
// (SHIP-148) says what the role below actually permits.
//
// # Nothing here has a counterpart in internal/identity, and that is the ticket
//
// The *Done when* is "admin sign-in is independent and cannot be reached with a user token". The
// most reliable way to hold that is for the two systems to have no type, no table and no key
// material in common — so [Administrator] is not a `User`, [Session] is not a `device_sessions`
// row, and neither can be built from the other because neither package can see the other. `admin`
// may not import `identity`, and the boundary lint refuses it.
//
// What the two *do* share is `internal/passwords`, which is infrastructure both sit on. That is
// the opposite of a shared credential: it is one argon2id implementation rather than two that
// agree by comment (SHIP-15r, Docs/10 §3.4).
//
// The blank line below keeps this a file note rather than a second package comment.

package admin

import (
	"slices"
	"time"

	"github.com/google/uuid"
)

// Role is the named bundle of permissions an administrator holds.
//
// A name in the database and a set in Go. The reasoning for the split is on [Permission]
// (SHIP-148): what a role grants is a security control read on every request, not data an operator
// edits, so it lives in a reviewable Go table rather than in a grant table somebody can UPDATE.
//
// The three below are ordered **least privileged first**, which is the order 000801's
// `ck_admin_users_role` lists them in and the order [Roles] holds. That is not decoration: it is
// how a reader tells at a glance which end of the list the default is at.
type Role string

const (
	// RoleSupport is the minimum, and it is what an account with no role stated becomes.
	//
	// Read-only across the console: search users and jobs, read the queues, read a dispute, read
	// the audit trail. Everything it holds is a permission to *look*, which is the whole of
	// Docs/01 §4.6's first two capabilities and none of the rest.
	//
	// 000801 makes this the column default for the reason Docs/04 §9 asks for least-privilege
	// administrative access: an INSERT that forgets the role should produce the account that can
	// do least, never the one that can do most.
	RoleSupport Role = "support"

	// RoleModerator is support plus the acting permissions of Docs/04 §6's step 4: unpublish a
	// job, restrict an account, record a verification decision, resolve a dispute, write an
	// internal note.
	//
	// It deliberately cannot manage administrators. Somebody who can create an administrator can
	// create one with any role, which makes every other boundary in this file advisory.
	RoleModerator Role = "moderator"

	// RoleOwner is moderator plus `admins.manage`.
	//
	// The role that can grant roles, and therefore the one there should be very few of. It is not
	// "all permissions" by construction — it holds a listed set like the other two, so a
	// permission added later reaches it only by being written into [rolePermissions].
	RoleOwner Role = "owner"
)

// Roles is every role, least privileged first.
//
// Ordered rather than a set because the order is meaningful (see [Role]) and because it is what
// TestAdminRoleConstraintMatchesTheGoConstants compares against `ck_admin_users_role`. Docs/10 §3.4
// requires that pairing in both directions: a role the database accepts and Go has no bundle for is
// an administrator whose permissions are undefined, and one Go knows and the database refuses is a
// role nobody can be given.
var Roles = []Role{RoleSupport, RoleModerator, RoleOwner}

// Valid reports whether r is one of [Roles].
func (r Role) Valid() bool { return slices.Contains(Roles, r) }

// String is the stored form, which is also the wire form.
//
// No translation table, for the reason delivery.ProofExceptionReason gives about its own: no
// document fixes these strings, so a second spelling would be a mapping with nothing on the other
// side of it.
func (r Role) String() string { return string(r) }

// Status is whether an administrator account may be used.
//
// Two values rather than users' four. An administrator is not verified, is not restricted, and has
// no commercial standing — the only question is whether the account is still in service.
type Status string

const (
	// StatusActive is an account in service.
	StatusActive Status = "active"

	// StatusDisabled is an account that may no longer sign in.
	//
	// Disabled rather than deleted, because 000003's audit entries name this account for ever and
	// Docs/05 §3.1 keeps that history after its subject is gone.
	//
	// Disabling has to end the live sessions as well, and it does so without a second write:
	// [Authenticator.Resolve] reads the account on every request and refuses one that is not
	// active. An account that cannot sign in and can still act is not disabled, and a credential
	// read from a row is what makes that true at once rather than at the next expiry.
	StatusDisabled Status = "disabled"
)

// Statuses is every status.
//
// Paired with `ck_admin_users_status` by TestAdminStatusConstraintMatchesTheGoConstants, per
// Docs/10 §3.4.
var Statuses = []Status{StatusActive, StatusDisabled}

// Valid reports whether s is one of [Statuses].
func (s Status) Valid() bool { return slices.Contains(Statuses, s) }

// String is the stored form.
func (s Status) String() string { return string(s) }

// Administrator is one row of `admin_users`, without its credential.
//
// **There is no password hash field and there will not be one.** The hash is read inside
// credentials.go, compared, and discarded; a struct carrying it would travel into a handler, into a
// log record built from `%+v`, and eventually into a response. `identity.User` takes the same
// position and for the same reason.
type Administrator struct {
	ID    uuid.UUID
	Email string
	Name  string

	Role   Role
	Status Status

	CreatedAt time.Time
	UpdatedAt time.Time
}

// CanSignIn reports whether this account may start a session.
//
// A method rather than a comparison at each call site, so that a third status — should one ever be
// argued for — is one edit rather than a search for every `== StatusActive`. `identity.User` has
// the same method for the same reason.
func (a Administrator) CanSignIn() bool { return a.Status == StatusActive }

// Session is one row of `admin_sessions`, without its credential.
//
// The token is never here either: what is stored is a hash of it, and what is returned to the
// caller at sign-in is the token itself, once, from [Credentials.SignIn]. A session that could be
// read back into a usable credential would make every support query and every backup a way in.
type Session struct {
	ID      uuid.UUID
	AdminID uuid.UUID

	// IdleExpiresAt slides as the console is used. AbsoluteExpiresAt does not.
	//
	// Two fields rather than one because they answer different questions and lapse for different
	// reasons: the first says "nobody has been here for a while", the second says "this session
	// has been open long enough". [Session.ExpiresAt] is the one a client is told about.
	IdleExpiresAt     time.Time
	AbsoluteExpiresAt time.Time

	LastUsedAt time.Time

	// RevokedAt is zero while the session is live.
	RevokedAt time.Time

	CreatedAt time.Time
}

// ExpiresAt is when this session stops working, whichever limit comes first.
//
// The value a client is given, because a console that displayed the idle expiry alone would promise
// a session it may not get: an administrator two minutes from the absolute cap does not have
// another thirty minutes of idle window however busy they are. 000801's
// `ck_admin_sessions_idle_within_absolute` makes the two agree in the database; this makes them
// agree in the answer.
func (s Session) ExpiresAt() time.Time {
	if s.AbsoluteExpiresAt.Before(s.IdleExpiresAt) {
		return s.AbsoluteExpiresAt
	}
	return s.IdleExpiresAt
}

// Live reports whether the session may still be presented at the given instant.
//
// Revoked, idle out, or capped out — three ways to be finished and one answer, because a caller
// deciding whether to serve a request has no use for the distinction. [Authenticator.Resolve] keeps
// the distinction for the answer it writes, where an administrator whose session timed out is told
// something different from one whose session was signed out from another tab.
func (s Session) Live(now time.Time) bool {
	return s.RevokedAt.IsZero() && now.Before(s.ExpiresAt())
}

File: apps/admin/lib/administrator.ts
import { cookies } from "next/headers";

import { SESSION_COOKIE } from "@/lib/session";
import { isJSON, platform } from "@/lib/upstream";

/**
 * Who is signed in, resolved on the server on every render (SHIP-188a).
 *
 * # It is a server module and cannot become anything else
 *
 * It imports `next/headers`, which Next refuses to bundle into a client component — importing this
 * file from a `"use client"` module is a build failure rather than a review comment. That is the
 * structural half of the guarantee `lib/session.ts` makes: the credential cannot reach the browser's
 * JavaScript because the only code that reads it cannot run there. `lib/surface.test.ts` holds the
 * other half, over the files a build would not notice.
 *
 * # There is no route handler for this, deliberately
 *
 * Sign-in and sign-out need one because a form in a browser has to reach them. Nothing in the
 * browser needs to ask who is signed in — the server already knows before it renders, and a route
 * handler here would be a browser-reachable endpoint added for nobody. The panel's rule is
 * therefore one file per upstream endpoint rather than the driver portal's stricter "only a route
 * handler may name one", and `lib/surface.test.ts` asserts the count either way: this file names
 * exactly one platform path and it is the one below.
 *
 * # Every render costs one call, and that is the design rather than an oversight
 *
 * `internal/admin/adminauth.go` re-reads `admin_sessions` on every request precisely so that
 * revocation and a role change take effect at once. A panel that cached the answer would reintroduce
 * the staleness the platform pays an indexed read per request to avoid — an administrator whose
 * permissions were cut at 2am would go on being shown the screens until something evicted a cache.
 */

/** An administrator, exactly as `GET /v1/admin/me` serves them. */
export interface Administrator {
  id: string;
  email: string;
  name: string;
  role: string;

  /**
   * What the role holds, expanded by the platform (SHIP-148).
   *
   * **Read to hide and disable, never to decide.** `Docs/07` §3 is a rule about every client: the
   * client may hide, the platform rules. An administrator who edited this array in a response would
   * find every endpoint refusing them exactly as before, because the check reads the role out of
   * the session row on each request.
   */
  permissions: string[];

  created_at: string;
}

/** The session the browser is holding, as much of it as a client has any use for. */
export interface AdminSession {
  id: string;
  expires_at: string;
}

/**
 * The three answers, and the panel renders a different screen for each.
 *
 * **`unavailable` is separate from `signed-out` on purpose.** Collapsing them would send an
 * administrator to a sign-in form during an outage, where their correct password would fail for a
 * reason the form cannot express — and the panel would be telling them their credential is bad when
 * what is bad is the platform. That is the same distinction `Docs/09`'s SHIP-193 row insists on
 * between a vendor's "no results" and its "request denied", for the same reason: a refusal read as
 * an absence sends everybody looking in the wrong place.
 */
export type Viewer =
  | { state: "signed-in"; administrator: Administrator; session: AdminSession }
  | { state: "signed-out" }
  | { state: "unavailable" };

interface MeResponse {
  administrator: Administrator;
  session: AdminSession;
}

/**
 * Resolve the cookie this browser presented into an administrator, or say why not.
 *
 * The cookie is read here and the credential goes out in a header. A header reaches no access log;
 * a request line does, which is why nothing about the credential is ever in a path or a query
 * string on any hop.
 */
export async function viewer(): Promise<Viewer> {
  const token = (await cookies()).get(SESSION_COOKIE)?.value;
  if (token === undefined || token === "") return { state: "signed-out" };

  let upstream: Response;
  try {
    upstream = await fetch(`${platform()}/v1/admin/me`, {
      method: "GET",
      headers: {
        Accept: "application/json",
        Authorization: `Bearer ${token}`, // spelling:ok — RFC 9110 header name
      },
      cache: "no-store",
    });
  } catch {
    return { state: "unavailable" };
  }

  // 401 is the platform saying this credential is not good: expired, signed out, or the account
  // disabled since it was issued. All three are "sign in again", which is the only thing an
  // administrator can do about any of them, so the panel does not distinguish what it cannot act on.
  if (upstream.status === 401) return { state: "signed-out" };
  if (!upstream.ok || !isJSON(upstream)) return { state: "unavailable" };

  let body: MeResponse;
  try {
    body = (await upstream.json()) as MeResponse;
  } catch {
    return { state: "unavailable" };
  }

  if (typeof body.administrator?.id !== "string") return { state: "unavailable" };

  return {
    state: "signed-in",
    administrator: body.administrator,
    session: body.session,
  };
}
